REGRESSION(r255164) [PlayStation] WTFReportBacktrace tries to print backtrace even when backtrace cannot be obtained and crashes

tomoki opened the pull request:
#### a4279526dfa54a830c0dce1130887778043b84f4
<pre>
REGRESSION(r255164) [PlayStation] WTFReportBacktrace tries to print backtrace even when backtrace cannot be obtained and crashes

<a href="https://bugs.webkit.org/show_bug.cgi?id=261497">https://bugs.webkit.org/show_bug.cgi?id=261497</a>

Reviewed by Kimmo Kinnunen.

WTFPrintBacktraceWithPrefixAndPrintStream should wrap the size parameter to 0 if it is negative not to print stack.

Currently, when !HAVE(BACKTRACE) &amp;&amp; !OS(WINDOWS),
- WTFGetBacktrace(samples, &amp;frames) makes frames = 0
- WTFReportBacktraceWithPrefixAndPrintStream passes -2 (=frames-framesToSkip) to WTFPrintBacktraceWithPrefixAndPrintStream.
- WTFPrintBacktraceWithPrefixAndPrintStream static_cast -2 to size_t, which can overflow and makes large number.
- It possibly tries to print the large stack and eventually crashes.

* Source/WTF/wtf/Assertions.cpp:
(WTFPrintBacktraceWithPrefixAndPrintStream): Wraps negative size to 0.
(WTFReportBacktraceWithPrefixAndPrintStream): Check the frame size and print &quot;no stacktrace available&quot; if the size is not enough.
(WTFReportBacktrace): Ditto.

Canonical link: <a href="https://commits.webkit.org/268121@main">https://commits.webkit.org/268121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfac97c3e2365f79ad19a780f8c1bdd5ed8b004a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17550 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19371 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19120 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21496 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16342 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17089 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23530 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16266 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17369 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17261 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21433 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18073 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15153 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22121 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16920 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5393 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4455 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21286 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23367 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17696 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5255 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple reviewed (commented):
Would it be more obvious to do something sane in the caller when the backtrace fails?
   ```
 WTFGetBacktrace(samples, &frames);
    if (frames < 2) {
         CrashLogPrintStream{}.print("%sno stacktrace available", prefix);
         return;
    }
    WTFPrintBacktraceWithPrefixAndPrintStream(out, samples + framesToSkip, frames - framesToSkip, prefix);
```

fujii committed:
REGRESSION(268090@main): WCBackingStore.h(61,31): warning: use 'template' keyword to treat 'decode' as a dependent template name
https://bugs.webkit.org/show_bug.cgi?id=261721

Unreviewed build fix for clang-cl Windows build.

* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.h:
(WebKit::WCBackingStore::decode): Use `template` keyword.

Canonical link: https://commits.webkit.org/268109@main

mattwoodrow committed:
css/filter-effects/repaint-added-backdrop-filter.html fails
https://bugs.webkit.org/show_bug.cgi?id=261713
<rdar://115701004>

Unreviewed, fix a prefixed property conversion that the import script missed.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/repaint-added-backdrop-filter.html:

Canonical link: https://commits.webkit.org/268110@main

nikolaszimmermann committed:
[LBSE] Enable accelerated transform animations for SVG renderers
https://bugs.webkit.org/show_bug.cgi?id=258122

Reviewed by Rob Buis.

Accelerated transform animations were not active for SVG renderers
in LBSE. RenderLayerBacking::startAnimation() only used accelerated
mode for RenderBoxModelObject derived renderers -- extend that to
all SVG layer awarerenderers. Forgot to upstream this half a year
ago :-(

Also make sure that 'transform-box' is respected, when calculating
the bounding box, against which e.g. 'transform-origin' percentage
values are resolved. This needs more testing to verify this outside
of SVG content in future -- need to coordinate with WPT.

Covered by existing tests, performance difference is significant.

Canonical link: https://commits.webkit.org/268111@main

carlosgcampos committed:
[GTK] build breaks with -DUSE_GBM=OFF: WebCore/DMABufFormat.h: No such file or directory
https://bugs.webkit.org/show_bug.cgi?id=261637

Reviewed by Michael Catanzaro.

* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::create):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:

Canonical link: https://commits.webkit.org/268112@main

Constellation committed:
[WebAnimation] Optimize updateCSSTransitions
https://bugs.webkit.org/show_bug.cgi?id=261715
rdar://115703255

Reviewed by Mark Lam.

WebAnimation's updateCSSTransitions is inefficient,

1. We are repeatedly copying AnimationProperty while it is held. This is appearing much in the profiler.
   Let's not copy them. We use `const AnimationProperty&`. This is OK since
   1.1. updateCSSTransitionsForStyleableAndProperty's owner is holding this ownership. Reference is always valid.
   1.2. Other change immediately copies or uses switchOn to dispatch. So this is fine.
2. Avoid repeated call of `styleable.hasRunningTransitionForProperty` etc. This is really costly function since
   it involves HashMap lookup. We reorganize updateCSSTransitionsForStyleableAndProperty not to do wasteful
   operations multiple times.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyBlendingContext::CSSPropertyBlendingContext):
(WebCore::CSSPropertyAnimation::blendProperty):
(WebCore::CSSPropertyAnimation::isPropertyAnimatable):
(WebCore::CSSPropertyAnimation::isPropertyAdditiveOrCumulative):
(WebCore::CSSPropertyAnimation::propertyRequiresBlendingForAccumulativeIteration):
(WebCore::CSSPropertyAnimation::animationOfPropertyIsAccelerated):
(WebCore::CSSPropertyAnimation::propertiesEqual):
(WebCore::CSSPropertyAnimation::canPropertyBeInterpolated):
* Source/WebCore/animation/CSSPropertyAnimation.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasCompletedTransitionForProperty const):
(WebCore::Element::hasRunningTransitionForProperty const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::keyframeEffectForElementAndProperty):
(WebCore::propertyInStyleMatchesValueForTransitionInMap):
(WebCore::transitionMatchesProperty):
(WebCore::updateCSSTransitionsForStyleableAndProperty):
(WebCore::Styleable::updateCSSTransitions const):
* Source/WebCore/style/Styleable.h:
(WebCore::Styleable::hasCompletedTransitionForProperty const):
(WebCore::Styleable::hasRunningTransitionForProperty const):

Canonical link: https://commits.webkit.org/268113@main

Constellation committed:
[WebCore] Do not lookup HashSet multiple times in PropertyCascade
https://bugs.webkit.org/show_bug.cgi?id=261722
rdar://115706784

Reviewed by Antti Koivisto.

Let's not lookup the same HashSet / HashMap multiple times.
And use HashSet<AtomString> instead of HashSet<String> since name is already atomized.

* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::set):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomProperties):
(WebCore::Style::Builder::applyCustomProperty):
(WebCore::Style::Builder::applyCustomPropertyImpl):
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderState.h:

Canonical link: https://commits.webkit.org/268114@main

tadeuzagallo committed:
[WGSL] Global variable rewriter is wrong for overrides
https://bugs.webkit.org/show_bug.cgi?id=261677
rdar://115655426

Reviewed by Dan Glastonbury.

When determining which global variables were used, the rewriter should skip the
path that deals with resources when visiting overrides, as they aren't resources.
The fix is trivial, I originally wrote a `break` where it should be `continue`

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::determineUsedGlobals):

Canonical link: https://commits.webkit.org/268115@main

tadeuzagallo committed:
[WGSL] Add matrix type aliases
https://bugs.webkit.org/show_bug.cgi?id=261675
rdar://115652287

Reviewed by Dan Glastonbury.

Add aliases for all matrix sizes such that matCxRf = matCxR<f32>

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):
* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/tests/valid/aliases.wgsl:

Canonical link: https://commits.webkit.org/268116@main

tadeuzagallo committed:
[WGSL] Add type declarations for built-in texture functions
https://bugs.webkit.org/show_bug.cgi?id=261682
rdar://115663474

Reviewed by Dan Glastonbury.

Add type declarations for all the built-in functions from section 16.7 of the spec[1]. The
only missing declarations now are the ones that use texture storage and texture depth, as
the first isn't support in type declarations yet and the latter isn't supported at all.

[1]: https://www.w3.org/TR/WGSL/#texture-builtin-functions

* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/268117@main

obyknovenius committed:
[WPE][GTK] Gardening `TestWebKitUserContentManager` `WebKit.MouseMoveAfterCrash` API test.

Unreviewed test gardening.

There is no such test.

* Tools/TestWebKitAPI/glib/TestExpectations.json:

Canonical link: https://commits.webkit.org/268118@main

tadeuzagallo committed:
[WGSL] Add code generation for compound assignment
https://bugs.webkit.org/show_bug.cgi?id=261679
rdar://115657490

Reviewed by Dan Glastonbury.

We were missing an entry for AST::CompoundAssignmentStatement in MetalFunctionWriter,
which meant the default visitor function was called, resulting in the left and right
hand sides of the assignment being serialized, but not the assignment operator.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: https://commits.webkit.org/268119@main

jyavenard committed:
Implement NativePromise
https://bugs.webkit.org/show_bug.cgi?id=254502
rdar://107546076

Reviewed by Youenn Fablet.

This is a WTF C++ native JS's like Promise, inspired by Gecko's MozPromise object

There are some significant differences from the gecko's implemtation,
both from a modus of operations and syntax usage in order to better fit
with WebKit, in particular:
- NativePromise can have a resolveType of `void`
- NativePromise takes a SerialFunctionDispatcher with ref/deref methods and
  is a template argument.
- There's typically no need to specify the return types of the resolve/reject callbacks.
  `auto` should work in most cases.

About NativePromise:

A promise manages an asynchronous request that may or may not be able to be fulfilled immediately.
When an API returns a promise, the consumer may attach callbacks to be invoked (asynchronously, on a specified thread)
when the request is either completed (resolved) or cannot be completed (rejected).

A NativePromise object is thread safe, and may be ->then()ed on any thread.
The then() call accepts either a resolve and reject callback, while whenSettled() accepts a resolveOrReject one.

NativePromise::then() and NativePromise::whenSettled() returns a NativePromise::Request object. This request can be either:
1- Converted back to a NativePromise which will be resolved or rejected once the resolve/reject callbacks are run.
  This new NativePromise can be then()ed again to chain multiple operations.
2- Be tracked using a NativePromiseRequest: this allows the caller to cancel the delivery of the resolve/reject result if it has not already occurred.
  (call to NativePromiseRequest::disconnect() must be done on the target thread to avoid thread safety issues).

When IsExclusive is true:
- The NativePromise performs run-time assertions that there is at most one call to either then(...) or chainTo(...).
- Move semantics are used when passing arguments
- The resolved or rejected object will be deleted on the target thread.
- The ResolveValueType and RejectValueType must have a move constructor if IsExclusive is true. Compilation will fail otherwise.
Otherwise:
- values are passed to the resolve/reject callbacks through either const references or pointers.
- the resolve or reject object will be deleted on the last SerialFunctionDispatcher that got used.

A typical workflow would be as follow:
If the work is to be done immediately:
From the producer side:
- Do the work
- return a resolved or rejected promise via NativePromise::createAndResolve or NativePromise::createAndReject
From the consumer side:
- call the method returning a promise
- then()/whenSettled() on the promise to set the actions to run once the promise has settled.

If the work is to be done at a later stage:
From the producer side:
- Allocate a NativePromise::Producer (via NativePromise::Producer::create() and return it to the consumer has a Ref<NativePromise>
- Do the work
- Once the work has been completed, either resolve or reject the NativePromise::Producer object.
From the consumer side:
- call the method returning a promise
- then() on the promise to set the actions to run once the promise has settled.

In either case (immediate or later resolution) using a NativePromiseRequest:
- track the promise
- cancel the delivery of the resolve/reject result and prevent callbacks to be run.

By disconnecting the NativePromiseRequest (via NativePromiseRequest::disconnect(), the then() callbacks will not be run.

Example:
```
        // You can mix & match promise types and chain them together.
        using MyPromise = NativePromise<int, int, true>;
        GenericPromise::Producer p(__func__);
        using MyPromise = NativePromise<int, int, true>;
        p->whenSettled(queue, __func__, [] (GenericPromise::Result result) {
            return MyPromise::createAndResolve(1, __func__);
        })->whenSettled(queue, __func__, [queue] (MyPromise::Result result) {
            static_assert(std::is_same_v<MyPromise::Result::value_type, int>, "The type received is the same as the last promise returned");
            EXPECT_TRUE(result.has_value());
            EXPECT_EQ(result.value(), 1);
            queue->beginShutdown();
        });
        p.resolve(__func__);
```

API tests added (re-used from Gecko's source code and adapted for NativePromise)

Canonical link: https://commits.webkit.org/268120@main

tomoki committed:
REGRESSION(r255164) [PlayStation] WTFReportBacktrace tries to print backtrace even when backtrace cannot be obtained and crashes

https://bugs.webkit.org/show_bug.cgi?id=261497

Reviewed by Kimmo Kinnunen.

WTFPrintBacktraceWithPrefixAndPrintStream should wrap the size parameter to 0 if it is negative not to print stack.

Currently, when !HAVE(BACKTRACE) && !OS(WINDOWS),
- WTFGetBacktrace(samples, &frames) makes frames = 0
- WTFReportBacktraceWithPrefixAndPrintStream passes -2 (=frames-framesToSkip) to WTFPrintBacktraceWithPrefixAndPrintStream.
- WTFPrintBacktraceWithPrefixAndPrintStream static_cast -2 to size_t, which can overflow and makes large number.
- It possibly tries to print the large stack and eventually crashes.

* Source/WTF/wtf/Assertions.cpp:
(WTFPrintBacktraceWithPrefixAndPrintStream): Wraps negative size to 0.
(WTFReportBacktraceWithPrefixAndPrintStream): Check the frame size and print "no stacktrace available" if the size is not enough.
(WTFReportBacktrace): Ditto.

Canonical link: https://commits.webkit.org/268121@main
